packages/solid-query/src/createBaseQuery.ts

```diff
@@ -35,46 +42,99 @@ export function createBaseQuery<
   const queryClient = createMemo(() =>
     useQueryClient({ context: options().context }),
   )
-  const emptyData = Symbol('empty')
 
   const defaultedOptions = queryClient().defaultQueryOptions(options())
   defaultedOptions._optimisticResults = 'optimistic'
   const observer = new Observer(queryClient(), defaultedOptions)
 
+  const emptyData = Symbol('empty')
+
+  let resolver: (value: TData | undefined) => void
+  let queryResolver: (value: DehydratedState) => void
+
+  // This resource will be used in a Server environment to
+  // dehydrate the queryClient when the query is
+  // pre fetched on the server. It will always be undefined
+  // if an observer is mounted on client.
+  const [queryResource] = createResource<DehydratedState | undefined>(() => {
+    return new Promise((resolve) => {
+      if (!isServer) {
+        resolve(undefined)
+      }
+      queryResolver = resolve
+    })
+  })
+
+  // If queryResource is defined,
+  // This means that the query was fetched on the server!
+  // We need to hydrate the queryClient with the query from
+  // the server before we set up the observer and its results.
+  if (queryResource()) {
```

Comment:
Minor, but could use the native `onHydrated` option on `createResource` up on line 59? E.g.

```ts
createResource<DehydratedState | undefined>(
  () => {
    return new Promise((resolve) => {
      if (!isServer) {
        resolve(undefined)
      }
      queryResolver = resolve
    })
  },
  {
    onHydrated(_k, info) {
      // We need to hydrate the queryClient with the query from
      // the server before we set up the observer and its results.
      hydrate(queryClient(), info.value)
    },
  },
)
```

Reply:
Scratch that - putting it here results in the `hydrate` call not happening until after the `observer.subscribe`, which is not what we want.

Reply:
Yeah I initially had the hydration on the `onHydrate` callback but seems to run as an effect after flushing the function so had to do this hacky resolver way